### PR TITLE
Add opt-in chpl compiler driver mode

### PR DESCRIPTION
**Pull request by riftEmber**
Adds compiler driver capability to `chpl` which can be opted-in to with `--compiler-driver`.

#### Summary
When `chpl` is invoked with `--compiler-driver`, it will perform initialization such as argument parsing, then re-invoke itself in two separate subprocesses in sequence: "phase one", which runs through the `codegen` pass, and "phase two", which picks up at the `makeBinary` pass.

#### Re-invocation process

- These re-invocations are full invocations of `chpl`, but with flags for controlling which phase of compilation is performed. The presence of the `--compiler-driver` flag, as well as the internal `--driver-phase-one` and `--driver-phase-two` flags, are used to skip sections of compiler work as necessary and/or possible. In particular, in phase one passes stop after the `codegen` pass, and in phase two passes are skipped up until the `makeBinary` pass. Effectively, `chpl --compiler-driver` is a thin wrapper which invokes separate executables for the phases.
- The parent process (initial invocation) is responsible for managing resources shared between the phases, such as a tmp dir handle. More information used by phase one is now saved to the tmp dir, for later use by phase two.
- Phase two is skipped entirely if it is not required, such as when `--parse-only` is set, or for GPU compilation at the moment (see "Work done in each phase").

##### Use with debuggers
When using `--compiler-driver` with a debugger flag (`--gdb` or `--lldb`), the `--driver-debug-phase` flag can be set to `1`, `2`, or `all` to determine which phases will be run in the debugger. Behavior with each option:
- `1`: Default if `--driver-debug-phase` is not set. Runs only phase one in the debugger, then exits without running phase two. This is because in most cases when debugging something in phase one, it will exit without fulfilling the duties phase two will expect to be done, leading to non-useful additional errors. A warning is emitted after debugger exit in this case to avoid misleading developers.
- `2`: Runs phase one normally without debugger, then starts up the debugger for phase two.
- `all`: Runs phase one in the debugger, then after that exits starts up the debugger again for phase two.

When running with a debugger, the driver first re-invokes `chpl` for the appropriate phase, then that sub-invocation re-invokes itself inside the debugger in turn (depending on `--driver-debug-phase`). If one needs to debug the behavior of `chpl` outside of the individual phases (i.e. debug the driver), using the debugger flags will not give the desired result; invoke `chpl` in `gdb`/`lldb` directly without use of the flag.

#### Work done in each phase
The division of work between the phases differs based on which backend is being used:
- C
	- 1: Does everything through generating C code + Makefile output.
	- 2: Invokes `make` on generated sources.
- LLVM
	- 1: Does everything through generating the LLVM `Module`, as well as all LLVM optimizations.
	- 2: Generates `.o` file from LLVM `Module`, compiles C source input files, and invokes LLVM linking.
- GPU:
	- 1: Does _everything_. This is because in GPU compilation, a child process is forked off inside the `codegen` pass to complete all necessary steps including invocation of the `makeBinary` function. This was the case before this PR, and could likely be changed in future.
	- 2: Is skipped entirely. Note the `makeBinary` _function_ still runs, just in the previous phase.

#### Future work
We would like compiler driver mode to be on by default in the future, but it is not yet robust enough. Required future work to make this happen is tracked in https://github.com/Cray/chapel-private/issues/5373.

------

Contributes to https://github.com/chapel-lang/chapel/issues/20871, https://github.com/Cray/chapel-private/issues/1291, and https://github.com/chapel-lang/chapel/issues/20845.

[reviewer info placeholder]

Testing:
- [x] local paratest
- [x] gasnet paratest
- [x] C backend paratest
- [x] GPU tests
- [x] `--verify` paratest
- [x] all of the above without `--compiler-driver`

### CODE REVIEW

**Comment by riftEmber** [comment]
Current status: working locally with C backend

File: test/compflags/ferguson/ldflags-order.prediff
Comment (DanilaFe) [inline test/compflags/ferguson/ldflags-order.prediff:21]: Does this technically provide an arbitrary code execution path for people running ldflags-order test? That would just mean we don't want to allow CI to run by default with this test, if that were ever a thing.

**Review by DanilaFe — APPROVED** [review]
None of the comments here are showstoppers; feel free to address as many as you like before merging, but everything looks good.

I think I would like an automatic testing configuration of this after the release.

File: compiler/llvm/clangUtil.cpp
Comment (ShreyasKhandekar) [inline compiler/llvm/clangUtil.cpp:5086]: Does this need the same check to ensure the correct LLVM version that you have above for the non gpu codegen?

File: compiler/main/driver.cpp
Comment (ShreyasKhandekar) [inline compiler/main/driver.cpp:830]: This confused me. Is the intention to not do the make binary phase if we're running inside the debugger or do we just run the make binary outside of the debugger?

File: compiler/main/driver.cpp
Comment (ShreyasKhandekar) [inline compiler/main/driver.cpp:927]: Perhaps a simpler and more informative message would be:

`"--driver-debug-phase requires either 1, 2, or 'all' as input, but got: %s\n"`

File: test/gpu/native/compilerDriver.chpl
Comment (ShreyasKhandekar) [inline test/gpu/native/compilerDriver.chpl:5]: Could you put an `@assertOnGpu` here to make sure we're codegen-ing for GPUs correctly with the compiler driver?

**Review by ShreyasKhandekar — APPROVED** [review]
Really Cool PR!!! 🎊 
I have some very minor comments :)

File: compiler/codegen/codegen.cpp
Comment (riftEmber) [reply compiler/codegen/codegen.cpp]: Ah, looks like this function is invoked by `makeBinaryLLVM` (and I think has to be) so the save/restore has to stay. The above comment about never having to emit the error on the back end is probably still valid however, so I'll adjust for that.

File: compiler/codegen/codegen.cpp
Comment (riftEmber) [reply compiler/codegen/codegen.cpp]: Implemented

File: compiler/include/driver.h
Comment (riftEmber) [reply compiler/include/driver.h:280]: I believe this was meant as "might not", and I removed the unneeded logic

File: compiler/include/files.h
Comment (riftEmber) [reply compiler/include/files.h:43]: Yes it is, but not something I think I should change in this PR. Do you know if the desire to remove this is tracked elsewhere? (It existed already, but was just not exported from `files.h` before)

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:4725]: Upon review I believe this call to the cleanup function is unnecessary. These are the per-function optimization managers which are not used in `llvmRunOptimizations`. Indeed, this cleanup function is called slightly before `llvmRunOptimizations` is in `finishCodegenLLVM`. I think that's the only call to it we need.

File: test/compflags/driver/debugger/declint-lldb.good
Comment (riftEmber) [reply test/compflags/driver/debugger/declint-lldb.good]: The OP was incorrect about this and I've updated it

File: compiler/include/files.h
Comment (riftEmber) [reply compiler/include/files.h:43]: Added a TODO here

Comment (DanilaFe) [reply compiler/include/files.h:43]: A "compiler"-tagged issue might be nice; I suspect no-one goes through the C++ source code comments to garbage-collect them.

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:5308]: It appears the only way for this error to be encountered is if the user does specify Chapel symbols to be disassembled, but we don't find them so we don't create cname mappings for them. Otherwise, the argument parser would complain to them about missing required arguments. I'll change the wording of this error to say we couldn't find any of the symbols that were requested to be disassembled, and then leave it to https://github.com/Cray/chapel-private/issues/5387 to emit this error earlier and tell the user what symbols couldn't be found.

File: compiler/include/files.h
Comment (riftEmber) [reply compiler/include/files.h:43]: Created https://github.com/Cray/chapel-private/issues/5390

File: compiler/util/files.cpp
Comment (riftEmber) [reply compiler/util/files.cpp:136]: Recorded in future work

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:5201]: Recorded in future work

File: compiler/include/files.h
Comment (riftEmber) [reply compiler/include/files.h:43]: (and recorded in future work issue)

File: compiler/util/files.cpp
Comment (riftEmber) [reply compiler/util/files.cpp:219]: Recorded in future work

File: frontend/lib/framework/Context.cpp
Comment (riftEmber) [reply frontend/lib/framework/Context.cpp:69]: Recorded in future work

**Comment by riftEmber** [comment]
@DanilaFe I have noted down all your comments I don't plan to get to on this PR in https://github.com/Cray/chapel-private/issues/5373, individually commented on each to say so and marked them resolved -- kind of verbose but just making sure I don't drop anything. The other comments I haven't marked as resolved I plan to address before merging.

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:4857]: It is, but for this PR I'm trying to stick with writing it explicitly as that's the preference Daniel expressed above. I think it's more commonly written explicitly in our compiler code.

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:4867]: Probably, but I would like to leave it in if it doesn't bother you too much as it was there before and I want to minimize incidental changes.

File: compiler/llvm/clangUtil.cpp
Comment (riftEmber) [reply compiler/llvm/clangUtil.cpp:5086]: No because the above version checks if we have LLVM > 6, and the lowest version we support is 11 so we are allowed to assume that's true. That means the above version check is also unnecessary -- we have a lot of unnecessary LLVM version checks to clean up at some point, and I would rather not do this one instance here.

File: compiler/main/driver.cpp
Comment (riftEmber) [reply compiler/main/driver.cpp]: The intention is to not run the make-binary phase ("phase 2" now) at all if we only ran the compilation phase ("phase 1") in the debugger. I wrote about this incorrectly in the OP and have since fixed it to reflect my intention and what the code does.

File: test/gpu/native/compilerDriver.chpl
Comment (riftEmber) [reply test/gpu/native/compilerDriver.chpl:5]: Good idea, will do

File: compiler/main/driver.cpp
Comment (riftEmber) [reply compiler/main/driver.cpp:918]: I like that phrasing better, will do

File: compiler/include/driver.h
Comment (riftEmber) [reply compiler/include/driver.h:280]: Hm, unfortunately looks like this is needed for generating the code for the `--about` exec flag, in `genConfigGlobalsAndAbout`, which occurs during code generation and probably couldn't be done by the driver itself. I've put it back.

File: test/compflags/ferguson/ldflags-order.prediff
Comment (riftEmber) [reply test/compflags/ferguson/ldflags-order.prediff]: I believe I've fixed this now, but am not experienced enough in Bash to be sure it's safe to run on untrusted input. However, I think that's alright as we'd have a lot of other precautions to worry about if we were going to allow CI to run our test suite; there are surely other `.prediff`s with similar issues, and in any case a user could PR changes to the `.prediff`s themselves to execute arbitrary code if we weren't careful.

**Pull request merged**
